### PR TITLE
introduce portfolio calculator factory 

### PR DESCRIPTION
**Pull request by x1c0**
this will allow different types of performance metrics, for now just TWR is working as before

### CODE REVIEW

File: apps/api/src/app/portfolio/calculator/twr/portfolio-calculator.ts [hunk 1]
```diff
@@ -1,663 +1,30 @@
...
-      };
-    }
+import { cloneDeep, first, last, sortBy } from 'lodash';
 
-    const currencies: { [symbol: string]: string } = {};
-    const dataGatheringItems: IDataGatheringItem[] = [];
-    let dates: Date[] = [];
-    let firstIndex = transactionPoints.length;
-    let firstTransactionPoint: TransactionPoint = null;
+import { CurrentPositions } from '../../interfaces/current-positions.interface';
```
Comment (dtslvr) [inline apps/api/src/app/portfolio/calculator/twr/portfolio-calculator.ts:20]: Can you keep the original import from above?

File: apps/api/src/app/portfolio/calculator/twr/portfolio-calculator.ts [hunk 2]
```diff
@@ -1,663 +1,30 @@
...
+import { PortfolioOrderItem } from '../../interfaces/portfolio-calculator.interface';
```
Comment (dtslvr) [inline apps/api/src/app/portfolio/calculator/twr/portfolio-calculator.ts:21]: Can you keep the original import from above? My hope is that the diff looks a little bit more readable.

File: apps/api/src/app/portfolio/calculator/twr/portfolio-calculator.ts [hunk 3]
```diff
@@ -754,126 +121,19 @@ export class PortfolioCalculator {
...
-    }
-  }
-
-  private getSymbolMetrics({
+  protected getSymbolMetrics({
     dataSource,
     end,
     exchangeRates,
-    isChartMode = false,
+    isChartMode,
```
Comment (dtslvr) [inline apps/api/src/app/portfolio/calculator/twr/portfolio-calculator.ts:133]: Why has this changed?

File: apps/api/src/app/portfolio/calculator/twr/portfolio-calculator.ts [hunk 4]
```diff
@@ -754,126 +121,19 @@ export class PortfolioCalculator {
...
     marketSymbolMap,
     start,
-    step = 1,
+    step,
```
Comment (dtslvr) [inline apps/api/src/app/portfolio/calculator/twr/portfolio-calculator.ts:136]: Why has this changed?

File: apps/api/src/app/portfolio/portfolio.module.ts [hunk 5]
```diff
@@ -42,7 +43,8 @@ import { RulesService } from './rules.service';
     AccountService,
     CurrentRateService,
     PortfolioService,
-    RulesService
+    RulesService,
+    PortfolioCalculatorFactory
```
Comment (dtslvr) [inline apps/api/src/app/portfolio/portfolio.module.ts:47]: Can you move up to keep the alphabetical order?

File: apps/api/src/app/portfolio/portfolio.service.ts [hunk 6]
```diff
@@ -106,7 +115,8 @@ export class PortfolioService {
     @Inject(REQUEST) private readonly request: RequestWithUser,
     private readonly rulesService: RulesService,
     private readonly symbolProfileService: SymbolProfileService,
-    private readonly userService: UserService
+    private readonly userService: UserService,
+    private readonly calculatorFactory: PortfolioCalculatorFactory
```
Comment (dtslvr) [inline apps/api/src/app/portfolio/portfolio.service.ts:119]: Can you move up to keep the alphabetical order?

File: apps/api/src/app/portfolio/portfolio.service.ts [hunk 7]
```diff
@@ -216,9 +226,11 @@ export class PortfolioService {
 
   public async getDividends({
     activities,
+    dateRange = 'max',
```
Comment (dtslvr) [inline apps/api/src/app/portfolio/portfolio.service.ts:229]: I think this is an artifact from the rebase, please remove

File: apps/api/src/app/portfolio/portfolio.service.ts [hunk 8]
```diff
@@ -216,9 +226,11 @@ export class PortfolioService {
...
     groupBy
   }: {
     activities: Activity[];
+    dateRange?: DateRange;
```
Comment (dtslvr) [inline apps/api/src/app/portfolio/portfolio.service.ts:233]: I think this is an artifact from the rebase, please remove

File: apps/api/src/app/portfolio/portfolio.service.ts [hunk 9]
```diff
@@ -232,7 +244,14 @@ export class PortfolioService {
       dividends = this.getDividendsByGroup({ dividends, groupBy });
     }
 
-    return dividends;
+    const startDate = this.getStartDate(
```
Comment (dtslvr) [inline apps/api/src/app/portfolio/portfolio.service.ts:247]: I think this block is an artifact from the rebase, please remove

File: apps/api/src/app/portfolio/portfolio.service.ts [hunk 10]
```diff
@@ -354,14 +372,13 @@ export class PortfolioService {
...
       activities,
-      currency: userCurrency,
-      currentRateService: this.currentRateService,
-      exchangeRateDataService: this.exchangeRateDataService
-    });
+      userCurrency
+    );
 
-    const { startDate } = getInterval(
+    const startDate = this.getStartDate(
```
Comment (dtslvr) [inline apps/api/src/app/portfolio/portfolio.service.ts:381]: I think this is an artifact from the rebase, please change

File: apps/api/src/app/portfolio/portfolio.service.ts [hunk 11]
```diff
@@ -946,10 +963,7 @@ export class PortfolioService {
     const userId = await this.getUserId(impersonationId, this.request.user.id);
     const user = await this.userService.user({ id: userId });
 
-    const { endDate, startDate } = getInterval(dateRange);
```
Comment (dtslvr) [inline apps/api/src/app/portfolio/portfolio.service.ts:949]: I think this is needed

File: apps/api/src/app/portfolio/portfolio.service.ts [hunk 12]
```diff
@@ -946,10 +963,7 @@ export class PortfolioService {
...
-
     const { activities } = await this.orderService.getOrders({
-      endDate,
```
Comment (dtslvr) [inline apps/api/src/app/portfolio/portfolio.service.ts:952]: I think this is needed

File: apps/api/src/app/portfolio/portfolio.service.ts [hunk 13]
```diff
@@ -963,17 +977,18 @@ export class PortfolioService {
...
+      PerformanceCalculationType.TWR,
       activities,
-      currency: this.request.user.Settings.settings.baseCurrency,
-      currentRateService: this.currentRateService,
-      exchangeRateDataService: this.exchangeRateDataService
-    });
+      this.request.user.Settings.settings.baseCurrency
+    );
 
-    const currentPositions = await portfolioCalculator.getCurrentPositions(
```
Comment (dtslvr) [inline apps/api/src/app/portfolio/portfolio.service.ts:973]: Please revert these changes

File: apps/api/src/app/portfolio/portfolio.service.ts [hunk 14]
```diff
@@ -1120,10 +1135,7 @@ export class PortfolioService {
       )
     );
 
-    const { endDate, startDate } = getInterval(dateRange);
```
Comment (dtslvr) [inline apps/api/src/app/portfolio/portfolio.service.ts:1123]: Please revert these changes

File: apps/api/src/app/portfolio/portfolio.service.ts [hunk 15]
```diff
@@ -1152,13 +1164,22 @@ export class PortfolioService {
...
+      PerformanceCalculationType.TWR,
       activities,
-      currency: userCurrency,
-      currentRateService: this.currentRateService,
-      exchangeRateDataService: this.exchangeRateDataService
-    });
+      userCurrency
+    );
+
+    const portfolioStart = min(
```
Comment (dtslvr) [inline apps/api/src/app/portfolio/portfolio.service.ts:1173]: Please revert these changes

File: apps/api/src/app/portfolio/portfolio.service.ts [hunk 16]
```diff
@@ -1172,7 +1193,7 @@ export class PortfolioService {
       netPerformancePercentageWithCurrencyEffect,
       netPerformanceWithCurrencyEffect,
       totalInvestment
-    } = await portfolioCalculator.getCurrentPositions(startDate, endDate);
+    } = await portfolioCalculator.getCurrentPositions(startDate);
```
Comment (dtslvr) [inline apps/api/src/app/portfolio/portfolio.service.ts:1196]: Please revert

File: apps/api/src/app/portfolio/portfolio.service.ts [hunk 17]
```diff
@@ -1428,11 +1448,11 @@ export class PortfolioService {
 
     userId = await this.getUserId(impersonationId, userId);
 
-    const { endDate, startDate } = getInterval(
+    const startDate = this.getStartDate(
```
Comment (dtslvr) [inline apps/api/src/app/portfolio/portfolio.service.ts:1451]: Please revert

File: apps/api/src/app/portfolio/portfolio.service.ts [hunk 18]
```diff
@@ -1597,6 +1617,52 @@ export class PortfolioService {
     };
   }
 
+  private getStartDate(aDateRange: DateRange, portfolioStart: Date) {
```
Comment (dtslvr) [inline apps/api/src/app/portfolio/portfolio.service.ts:1620]: I think this should be reverted

File: apps/api/src/app/portfolio/calculator/portfolio-calculator.factory.ts [hunk 19]
```diff
@@ -0,0 +1,47 @@
...
+
+@Injectable()
+export class PortfolioCalculatorFactory {
+  constructor(
+    private readonly currentRateService: CurrentRateService,
+    private readonly exchangeRateDataService: ExchangeRateDataService
+  ) {}
+
+  createCalculator(
+    calculationType: PerformanceCalculationType,
```
Comment (dtslvr) [inline apps/api/src/app/portfolio/calculator/portfolio-calculator.factory.ts:24]: Can you use an object, similar as it was before?

File: apps/api/src/app/portfolio/calculator/twr/portfolio-calculator-novn-buy-and-sell.spec.ts [hunk 20]
```diff
@@ -1,12 +1,19 @@
 import { Activity } from '@ghostfolio/api/app/order/interfaces/activities.interface';
 import { CurrentRateService } from '@ghostfolio/api/app/portfolio/current-rate.service';
-import { CurrentRateServiceMock } from '@ghostfolio/api/app/portfolio/current-rate.service.mock';
 import { ExchangeRateDataService } from '@ghostfolio/api/services/exchange-rate-data/exchange-rate-data.service';
 import { parseDate } from '@ghostfolio/common/helper';
 
 import { Big } from 'big.js';
 
-import { PortfolioCalculator } from './portfolio-calculator';
+import { CurrentRateServiceMock } from '../../current-rate.service.mock';
```
Comment (dtslvr) [inline apps/api/src/app/portfolio/calculator/twr/portfolio-calculator-novn-buy-and-sell.spec.ts:8]: Keep the import as it was before

**Review by dtslvr — CHANGES_REQUESTED** [review]
I have added some comments. It looks like something messed up during the rebase 😜

File: apps/api/src/app/portfolio/calculator/twr/portfolio-calculator.ts (hunk 3 above)
Comment (x1c0) [reply apps/api/src/app/portfolio/calculator/twr/portfolio-calculator.ts:133]: `A parameter initializer is only allowed in a function or constructor implementation.ts (2371)`

what I need to make sure is that when using this method `calculateOverallPerformance` on TWR this default value is set

File: apps/api/src/app/portfolio/calculator/twr/portfolio-calculator.ts (hunk 4 above)
Comment (x1c0) [reply apps/api/src/app/portfolio/calculator/twr/portfolio-calculator.ts:136]: `A parameter initializer is only allowed in a function or constructor implementation.ts (2371)`

what I need to make sure is that when using this method `calculateOverallPerformance` on TWR this default value is set

File: apps/api/src/app/portfolio/portfolio.service.ts (hunk 17 above)
Comment (x1c0) [reply apps/api/src/app/portfolio/portfolio.service.ts:1451]: no idea why this happened

**Review by dtslvr — APPROVED** [review]
